spec/namespaces.md

```diff
@@ -10,7 +10,7 @@ A *compilation_unit* defines the overall structure of a source file. A compilati
 
 ```antlr
 compilation_unit
-    : extern_alias_directive* using_directive* global_attributes? namespace_member_declaration*
+    : extern_alias_directive* using_directive* global_attributes? (single_line_namespace | (namespace_member_declaration*))
```

Comment:
Small nit, but I feel like this might read better if pulled out into `namespace_declaration?`, where `namespace_declaration` is an or between `single_line_namespace` and `namespace_member_declaration+`

Reply:
that's how i started, but that's actually a lot harder to deal with in the spec.  You now need the rules that ensure:

1. a single-line namespace precedes other members.
2. you can only have one single-line namespace
3. you can't have normal namespace follow a single-line namespace
4. a normal namespace can't contain a single-line namespace.

The grammar approach here really enforces all these rules succinctly.

With the above you really just have:

Either:
   1. a single-line namespace (which itself is easily defined)
   2. a normal namespace declaration

That's it :)

Reply:
I don't see how? I've just proposed breaking out what you have here:

```antlr
compilation_unit
    : extern_alias_directive* using_directive* global_attributes? namespace_declaration?

namespace_declaration
	: single_line_namespace
	| namespace_member_declaration+
```

That should be identical to what you have here, but more readable imo.

Reply:
I see what you mean now.  the problem is that `namespace_declaration` is already a production. We'd have to name this differently.  Perhaps `compilation_unit_body` ?  I would be fine with that name, or something else if you can think of another suitable one.

Reply:
`compilation_unit_body` sounds fine to me.

Reply:
Done :)